velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,241 @@
+    std::vector<const vector_size_t*> indices(args.size());
+
+    for (int i = 0; i < numInputArrays; i++) {
+      exec::LocalDecodedVector holder(context, *args[i].get(), rows);
+      decodedVectors[i] = holder.get();
+      baseVectors[i] = decodedVectors[i]->base()->as<ArrayVector>();
+      rawSizes[i] = baseVectors[i]->rawSizes();
+      rawOffsets[i] = baseVectors[i]->rawOffsets();
+      indices[i] = decodedVectors[i]->indices();
+    }
+
+    // Size of elements in result vector.
+    vector_size_t resultElementsSize = 0;
+
+    // Determine what the size of the resultant elements will be so we can
+    // reserve enough space.
+    auto getMaxArraySize =
+        [&](vector_size_t row) -> std::optional<vector_size_t> {
+      vector_size_t maxSize = 0;
+      for (int i = 0; i < numInputArrays; i++) {
+        if (decodedVectors[i]->isNullAt(row)) {
+          return std::nullopt;
+        } else {
+          maxSize = std::max(maxSize, rawSizes[i][indices[i][row]]);
+        }
+      }
+      return std::optional(maxSize);
+    };
+
+    rows.applyToSelected([&](auto row) {
+      auto maxArraySize = getMaxArraySize(row);
+      if (UNLIKELY(!maxArraySize.has_value())) {
+        resultElementsSize++;
+      } else {
+        resultElementsSize += maxArraySize.value();
+      }
+    });
+
+    // Now create individual result vectors for each input Array vector.
+
+    auto pool = context->pool();
+    std::vector<BufferPtr> resultIndices(numInputArrays);
+    std::vector<BufferPtr> resultElementNulls(numInputArrays);
+    std::vector<vector_size_t*> rawResultIndices(numInputArrays);
+    std::vector<uint64_t*> rawResultNulls(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultIndices[i] = allocateIndices(resultElementsSize, pool);
+      resultElementNulls[i] = AlignedBuffer::allocate<bool>(
+          resultElementsSize, pool, bits::kNotNull);
+      rawResultIndices[i] = resultIndices[i]->asMutable<vector_size_t>();
+      rawResultNulls[i] = resultElementNulls[i]->asMutable<uint64_t>();
+    }
+
+    // Final Row Vector
+    const auto finalArraySize = rows.size();
+    BufferPtr finalArrayIndices = allocateIndices(finalArraySize, pool);
+    BufferPtr finalArrayNulls =
+        AlignedBuffer::allocate<bool>(finalArraySize, pool, bits::kNotNull);
+    BufferPtr finalArrayOffsets = allocateOffsets(finalArraySize, pool);
+    BufferPtr finalArraySizes = allocateSizes(finalArraySize, pool);
+
+    auto rawFinalArrayNulls = finalArrayNulls->asMutable<uint64_t>();
+    auto rawFinalArrayOffsets = finalArrayOffsets->asMutable<vector_size_t>();
+    auto rawFinalArraySizes = finalArraySizes->asMutable<vector_size_t>();
+
+    // Create right offsets/indexes for the individual and final result arrays.
+    int rowIndex = 0;
+    int finalResultIndex = 0;
+    rows.applyToSelected([&](auto row) {
+      // Get the max size for that row.
+      auto maxRowSize = getMaxArraySize(row);
```

Comment:
Is it necessary to compute this twice? Can't we populate the offsets and sizes at the same time as computing total number of elements?

Reply:
The reason for 2 passes was so that I could allocate memory for the buffers etc. Not sure if we can dynamically allocate memory and resize Buffers and if thats a good idea.

Reply:
Makes sense. Perhaps, save the results of getMaxArraySize(row) in a buffer of rows.size() size to avoid recomputing.